### ISSUE TITLE
Support multiple backing services in customEnvVar / dataMapping

### ISSUE THREAD

**Issue by sbose78** [redhat-developer/service-binding-operator#396]
JIRA story https://issues.redhat.com/browse/APPSVC-481

Extracted out https://github.com/redhat-developer/service-binding-operator/issues/356#issuecomment-596108001

```
backingServiceSelectors:
    - group: postgres.dev
      kind: Service
      resourceRef: user-db
      version: v1beta1
      id: X
    - group: postgres.dev
      kind: Credentials
      resourceRef: admin-creds
      version: v1beta1
      id: Y
customEnvVar:
    - name: CONNECTION_URL
       value: {{  X.status.username }} / {{ Y.status.url }}
```

**Comment by sbose78** [comment]
We kinda decided on this, @ZhuangYuZY to make it easy for folks to write :) What do you think?

**Comment by sbose78** [comment]
@otaviof did you want `value: {{  X.status.username }} / {{ Y.status.url }` to look different?

**Comment by qibobo** [comment]
@sbose78 
To implement this, the retriever.cache will be changed from map[string]interface{} to map[string]map[string]interface{}?
For map[string]interface{}, the key is "status", or "spec" ....
For map[string]map[string]interface{}, the key are X, Y and the value are the same as current cache(map[string]interface{}).
Please correct me if I am wrong. Thanks.

**Comment by otaviof** [comment]
@sbose78, @qibobo regarding the data collection, lets sync with @isutton and @Avni-Sharma, they are currently busy with those changes.

**Comment by ZhuangYuZY** [comment]
@sbose78 Thank you. Will take some time for internal discussion to see if it works.

**Comment by sbose78** [comment]
FYI @arthurdm @navidsh tracking this separately here.

**Comment by ZhuangYuZY** [comment]
@sbose78 and @otaviof following our discussion on to-json requirement in #388 .

After discussion with team, we want to go with following way, which is using to-json in a template string as customEnvVar. I think it aligns with our discussion and operator design that extending customEnvVar to introduce a new to-json function, and also multiple backing services which identified by `id`.  Example is below. Let me know your comment. If agree, then let's discussion on implementation. 
There are three changes needed 1) multiple backing services no override on customEnvVar, 2) id for backing services, 3) to-json function based on id. My question is do you have any ongoing work for above ? I'd like understand and think how to collaborate to submit PR for 1), 2), and 3).
Let me know your comment. Thank you.

```
apiVersion: apps.openshift.io/v1alpha1
kind: ServiceBindingRequest
metadata:
  name: helloworld-binding
spec:
  applicationSelector:
    resourceRef: helloworld
    group: serving.knative.dev
    version: v1
    resource: services
  backingServiceSelectors:
    - group: ibmcloud.ibm.com
      version: v1alpha1
      kind: Binding
      resourceRef: coligo-cos-service-credential
      id: coligo-cos-service-credential
  customEnvVar:
    - name: VCAP_SERVICE
      value: {
          "cloud-object-storage": [
            {
              "credentials": {
                {{to_json coligo-cos-service-credential.status.secretName}} 
              },
              "name": "coligo-cos",
              "plan": "standard"                      
            }
          ]
        }
```

**Comment by qibobo** [comment]
@sbose78 and @otaviof
Is there any PR or work on it? If there is no, I have done some code change to implement the feature and can submit a PR for it.

**Comment by ZhuangYuZY** [comment]
@sbose78 and @otaviof,  we are preparing PR to support backingService id and to_json. Thank you.

**Comment by sbose78** [comment]
If you have local code changes, please free to open a WIP PR anyway to help us drive a nice discussion. I believe @isutton was working on this - I'll let him add more details. :)

**Comment by isutton** [comment]
I see the appeal of having a shortcut in there, but the shortcut might be
in the wrong place though.

For example, {{ X.status.username }} / {{ Y.status.url }} isn't far from {{
Credentials.admin-creds.status.username }} / {{ Service.user-db.status.url
}} for a one off usage (and to be fair, everything in k8s is verbose). If
the template expression is more complicated, like we've seen in some
examples, then the developer has the option to define template variables in
the template definition.

Thoughts?

**Comment by sbose78** [comment]
I'm fine with that Igor. Thanks for chipping with this perspective!

**Comment by sbose78** [comment]
How about supporting all of these:
- resourceRef.status.url ( all resourceRefs are unique, don't bother specifying the whole thing )
- kind.resourceRef.status.... 
-  group.kind.resourceRef.status...
- version.group.kind.resourceRef.status....

Supporting all of the above ensures that funny edge cases where the only thing unique is the API group ( for example) is also covered :-)

Note we support spec, status and data blocks in CRs/resources.

**Comment by ZhuangYuZY** [comment]
Hi, @sbose78 and @isutton , either `id` or version/group/kind/resourceRef as variable in template can satisfy the requirement from user perspective.

But from controller implementation perspective, I think it will be complicated to support all of above, resourceRef.status, kind.resourceRef.status, group.kind.resourceRef.status, version.group.kind.resourceRef.status. It can not be done by a simple map. We have to design new data structure to support all above.

My suggestion:
1. using `kind.resourceRef` as key of backing service instances map if not `id`. I think I can cover 80% cases as a unique identifier.
2. Use `id` as optional to cover 20% cases, if developer wants to use it.
From implementation, one simple map for backing service instances, key could either `kind.resourceRef` or `id`.

What do you think ? Thank you.

**Comment by arthurdm** [comment]
I think we need to look at from the angle of:  what is this `id` field used for?   

It's not to provide a cluster-wide ID, because if we were doing that then even `version/group/kind/resourceRef` doesn't cover the case where you could have same 4-tuple but in different namespaces (you would then need `namespace/version/group/kind/resourceRef`).

From my perspective this `id` is only used for two things:
* being able to reference a particular service within the customEnvVar / dataMapping field of the `ServiceBinding`
* for the application code to distinguish one binding from the other

So it doesn't need to be cluster wide, it just needs to simply be `ServiceBinding` wide - and most will have probably under 5 services, if that?    It can be very simple, like `mongo` or `postgresql`.

**Comment by sbose78** [comment]
@redhat-developer/service-binding Has this been addressed in #407 ?

**Comment by sbose78** [comment]
Done in #468
